src/jtag3.c

```diff
@@ -2823,25 +2817,19 @@ static void jtag3_print_parms(const PROGRAMMER *pgm, FILE *fp) {
   jtag3_print_parms1(pgm, "", fp);
 }
 
-static unsigned char jtag3_mtype(const PROGRAMMER *pgm, const AVRPART *p, unsigned long addr) {
-  if (p->prog_modes & PM_PDI) {
-    if (addr >= PDATA(pgm)->boot_start)
-      return MTYPE_BOOT_FLASH;
-    else
-      return MTYPE_FLASH;
-  } else {
-    return MTYPE_FLASH_PAGE;
-  }
+static unsigned char jtag3_mtype(const PROGRAMMER *pgm, const AVRPART *p, const AVRMEM *m,
+   unsigned long addr) {
+
+  return mem_is_boot(m)? MTYPE_BOOT_FLASH:
+    (mem_is_flash(m) && is_pdi(p) && addr >= PDATA(pgm)->boot_start)? MTYPE_BOOT_FLASH: MTYPE_FLASH;
```

Comment:
It looks like the write issue for classic parts is caused by this return statement. `MTYPE_FLASH_PAGE` will never be returned.

Reply:
No. My initial suggestion caused `-cpowerdebugger_jtag -patxmega256a3bu -T 'write flash 0 "Hello"'` to fail.

```c
return mem_is_boot(m)? MTYPE_BOOT_FLASH:
    (mem_is_flash(m) && is_pdi(p) && addr >= PDATA(pgm)->boot_start)? MTYPE_FLASH: MTYPE_FLASH_PAGE;
```

Can't we just revert back to:

```c
 if (p->prog_modes & PM_PDI) {
    if (addr >= PDATA(pgm)->boot_start)
      return MTYPE_BOOT_FLASH;
    else
      return MTYPE_FLASH;
  } else {
    return MTYPE_FLASH_PAGE;
  }
```

Since this works perfectly fine, and is IMO easier to read?

Reply:
On closer inspection, I think this return statement should be OK. Note the `addr < PDATA(pgm)->boot_start`:

```c
return mem_is_boot(m)? MTYPE_BOOT_FLASH:
    (mem_is_flash(m) && is_pdi(p) && addr < PDATA(pgm)->boot_start)? MTYPE_FLASH: MTYPE_FLASH_PAGE;
```

Reply:
I feel like this is some kind of Whack-a-Mole game. Even with the return statement fixed, The powerdebugger_jtag flash write progress suddenly freezes at 97% when flashing an ATxmega256A3BU.

I suggest fixing the issue mentioned in my previous post and then we can look into what's going on this time.

Reply:
>  Note the `addr < PDATA(pgm)->boot_start`

Great find. What was I thinking?

Reply:
I have redone the fix in a different way and force-pushed that. Now all moles should be hiding incl tpi, isp, pdi, updi...